sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/constraints.scala

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions
+
+import org.antlr.v4.runtime.ParserRuleContext
+
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.types.{DataType, StringType}
+
+trait TableConstraint {
+
+  /** Returns the name of the constraint */
+  def name: String
+
+  /** Returns the characteristics of the constraint (e.g., ENFORCED, RELY) */
+  def characteristic: ConstraintCharacteristic
+
+  /** Creates a new constraint with the given name
+   * @param name Constraint name
+   * @return New TableConstraint instance
+   */
+  def withName(name: String): TableConstraint
+
+  /** Creates a new constraint with the given characteristic
+   * @param c Constraint characteristic (ENFORCED, RELY)
+   * @param ctx Parser context for error reporting
+   * @return New TableConstraint instance
+   */
+  def withCharacteristic(c: ConstraintCharacteristic, ctx: ParserRuleContext): TableConstraint
+
+  /** Generates a constraint name if one is not provided
+   * @param tableName Name of the table containing this constraint
+   * @return TableConstraint with a generated name if original name was null/empty
+   */
+  def generateConstraintNameIfNeeded(tableName: String): TableConstraint = {
+    if (name == null || name.isEmpty) {
+      this.withName(name = generateConstraintName(tableName))
+    } else {
+      this
+    }
+  }
+
+  // Generate a constraint name based on the table name if the name is not specified
```

Comment:
What if the table is renamed later? It seems we can simplify this logic quite a bit without the need to include the table name. I understand it is done to distinguish constraints but I wonder if we can leverage the catalog/namespace/table identifier in that code rather than attempting to generate a unique enough name here?

Reply:
I do understand we replicate Postgres behavior here, but we don't guarantee there are no duplicates. Let's discuss a bit more on how we can implement things like `pg_constraint` table.

Reply:
> I wonder if we can leverage the catalog, namespace, table identifier in that code rather than attempting to generate a unique enough name here?

This is possible for PK & FK, but hard for check and unique constraints.

I am ok to include catalog and namespace. Probably in analyzer rule `ResolveTableSpec`. However, it is a bit tricky to get the table name from the current V2 `CreateTable` plan
```
case class CreateTable(
    name: LogicalPlan,
    columns: Seq[ColumnDefinition],
    partitioning: Seq[Transform],
    tableSpec: TableSpecBase,
    ignoreIfExists: Boolean)
```